---
title: "Tool for Immunoglobulin Genotype Elucidation via Rep-Seq (TIgGER)"
author: "Daniel Gadala-Maria"
date: "Last modified `r Sys.Date()`"
output: rmarkdown::html_vignette
vignette: >
  %\VignetteIndexEntry{Vignette Title}
  %\VignetteEngine{knitr::rmarkdown}
  %\usepackage[utf8]{inputenc}
---


Introduction
--------------------------------------------------------------------------------
Repertoire-Sequencing (Rep-Seq) data, derived from the cDNA of B cells'
transcribed immunoglobulin (Ig) genes, is currently the subject of much study.
A key step in analyzing these data involves assigning the closest known V(D)J 
germline alleles to the (often somatically mutated) sample sequences using a
tool such as IMGT/HighV-QUEST[[1]][1].However, if the sample utilizes alleles
not in the germline database used for alignment, this step will fail.
Additionally, this alignment has an associated error rate of ~5%, notably among
sequences carrying a large number of somatic mutations[[2]][2].

Here we provide a Tool for Immunoglobulin Genotype Elucidation via Rep-Seq
(TIgGER). TIgGER addresses these issues by inferring the set of Ig alleles
carried by an individual (including any novel alleles) and then using this set
of alleles to correct the initial assignments given to sample sequences by
existing tools.

Additional information is available at
[http://clip.med.yale.edu/tigger/](http://clip.med.yale.edu/tigger/) and a full
description of TIgGER is available in:

> Gadala-Maria *et al*. (2015) Automated analysis of high-throughput B cell
sequencing data reveals a high frequency of novel immunoglobulin V gene
segment alleles. *PNAS*.


Input
--------------------------------------------------------------------------------
TIgGER requires two main inputs:

1. Pre-processed Rep-Seq data
2. Database germline sequences

Rep-seq data input is a tab-delimited table, as is output by Change-O
[(http://clip.med.yale.edu/changeo/)](http://clip.med.yale.edu/changeo/).
Each row represents a unique functional sequence in this format, and columns
hold information including a unique ID, the IMGT-gapped nucleotide sequence, the
names of the best-aligning V(D)J alleles, the nucleotide length of each of the
alleles, and the length of the junction. The user should load this data into R
as a data fram. An example dataset is provided with the `tigger` package. It
contains unique sequences assigned to IGHV1 family genes isolated from
individual PGP1 (data used in Gadala-Maria *et al.*).

```{r, eval=TRUE, warning=FALSE}
library(tigger)

# Load example Rep-Seq data
data(pgp1_example)

```

The database of germline sequences should be provided in FASTA format with
sequences gapped according to the IMGT numbering scheme[[3]][3]. IGHV alleles
in the IMGT database (build 201408-4) are provided with this package. You may
read in your own fasta file using `readGermlineDb()`.
```{r, eval=TRUE, warning=FALSE}
# Load example germline database
data(germline_example)
```


Running TIgGER
--------------------------------------------------------------------------------
The function most users will wish to use is `runTigger()`. This function takes
the two inputs discussed above and can perform any combination of the following:

1. Infer the presence of novel IGHV alleles not in the germline database
2. Infer the individual's IGHV genotype
3. Correct the IGHV allele calls of the samples based on the IGHV genotype

These options are controlled by the logical arguments `find_novel`,
`find_genotype`, and `correct_calls`, which default to `TRUE`. Note, however,
that if `find_genotype` is `FALSE`, the provided germline database should
contain only the alleles carried by the individual from which the Rep-Seq data
was isolated.

```{r, eval=TRUE, warning=FALSE}
# Run tigger on the sample output
sample_output = runTigger(pgp1_example, germline_example)
```

[1]: http://www.imgt.org/IMGTindex/IMGTHighV-QUEST.html "Alamyar et al. (2010)"
[2]: http://www.ncbi.nlm.nih.gov/pubmed/20147303 "Munshaw and Kepler (2010)"
[3]: http://www.ncbi.nlm.nih.gov/pubmed/12477501 "Lefranc et al. (2003)"








































