**inferGenotypeBayesian** - *Infer a subject-specific genotype using a Bayesian approach*

Description
--------------------

`inferGenotypeBayesian` infers an subject's genotype applying a Bayesian framework with Dirichlet prior 
for the multinomial distribution. Up to four distinct alleles are allowed in an
individualâ€™s genotype. Four likelihood distributions were generated by 
fitting empirically three high coverage genotypes from three individuals (ref fludata?). A posterior 
probability is calculated for the four most common alleles. The certainty of
the highest probability model was calculated using a Bayes factor (the most likely model
divided by second-most likely model). The larger the Bayes factor (K), 
the greater the certainty in the model.


Usage
--------------------
```
inferGenotypeBayesian(clip_db, v_call = "V_CALL",
find_unmutated = TRUE, germline_db = NA, novel_df = NA,
priors = c(0.6, 0.4, 0.4, 0.35, 0.25, 0.25, 0.25, 0.25, 0.25))
```

Arguments
-------------------

clip_db
:   a `data.frame` containing V allele
calls from a single subject. If
`find_unmutated` is `TRUE`, then
the sample IMGT-gapped V(D)J sequence should

v_call
:   column in `clip_db` with V allele calls.
Default is `"V_CALL"`
be provided in a column `"SEQUENCE_IMGT"`

find_unmutated
:   if `TRUE`, use `germline_db` to
find which samples are unmutated. Not needed
if `allele_calls` only represent
unmutated samples.

germline_db
:   named vector of sequences containing the
germline sequences named in
`allele_calls`. Only required if
`find_unmutated` is `TRUE`.

novel_df
:   an optional `data.frame` of the type
novel returned by
[findNovelAlleles](findNovelAlleles.md) containing
germline sequences that will be utilized if
`find_unmutated` is `TRUE`. See
details.

priors
:   vector of priors for the multinomial 
distribution. The first two enteries represent
two allele priors, next three represent 
three allele priors and the last four, four 
allele priors. Each set of priors should sum to one.




Value
-------------------

A table of alleles denoting the genotype of the subject with the log10
of the likelihood of each model and the log10 of the Bayes factor.


Details
-------------------

Allele calls representing cases where multiple alleles have been
assigned to a single sample sequence are rare among unmutated
sequences but may result if nucleotides for certain positions are
not available. Calls containing multiple alleles are treated as
belonging to all groups. If `novel_df` is provided, all
sequences that are assigned to the same starting allele as any
novel germline allele will have the novel germline allele appended
to their assignent prior to searching for unmutated sequences.


Note
-------------------

This method works best with data derived from blood, where a large
portion of sequences are expected to be unmutated. Ideally, there
should be hundreds of allele calls per gene in the input.



Examples
-------------------

```R
# Infer the IGHV genotype, using only unmutated sequences, including any 
# novel alleles
inferGenotypeBayesian(SampleDb, find_unmutated=TRUE, germline_db=GermlineIGHV,
novel_df=SampleNovel)
```


```
        GENE     ALLELES         COUNTS TOTAL NOTE                KH                KD                KT                KQ
1    IGHV1-2       02,04        664,302   966                  -1000 -7.92846809405969 -139.556367176944 -313.583949130729
2    IGHV1-3          01            226   226       4.20089197988625 -45.2911957825576 -84.2865868763307 -128.991761853586
3    IGHV1-8 01,02_G234T        467,370   837                  -1000 -1.04759115960507 -102.524664723923 -247.193958844361
4   IGHV1-18          01           1005  1005      -3.76643736033536  -223.85293382607             -1000             -1000
5   IGHV1-24          01            105   105       4.75335701924247 -18.2407545518045 -36.3580822723628 -57.1281856909991
6   IGHV1-46          01            624   624      0.457455409315221 -136.193264784335 -243.861955237939             -1000
7   IGHV1-58       01,02          23,18    41      -20.3932114156223  3.60009261357983 -1.38512929425796 -8.47869574581951
8   IGHV1-69 01,04,06,02 515,469,280,15  1279                  -1000 -277.291087469703  3.55051520054669 -143.380669247128
9 IGHV1-69-2          01             31    31       4.16107190423977 -2.62766579768837 -7.97659112471034 -14.1087168959268
             Kdiff
1 131.627899082884
2 49.4920877624439
3 101.477073564318
4 220.086496465735
5 22.9941115710469
6  136.65072019365
7  4.9852219078378
8 146.931184447674
9 6.78873770192814

```



See also
-------------------

[plotGenotype](plotGenotype.md) for a colorful visualization and
[genotypeFasta](genotypeFasta.md) to convert the genotype to nucleotide sequences.
See [inferGenotype](inferGenotype.md) to infer a subject-specific genotype using 
a frequency method



